storage/inkless/src/main/resources/db/migration/V8__Find_batches_function_with_limit.sql

```diff
@@ -0,0 +1,88 @@
+CREATE OR REPLACE FUNCTION find_batches_v2(
+    arg_requests find_batches_request_v1[],
+    fetch_max_bytes INT,
+    max_batches_per_partition INT DEFAULT 0
+)
+RETURNS SETOF find_batches_response_v1 LANGUAGE sql STABLE AS $$
+    WITH
+        requests AS (
+            SELECT
+                r.topic_id,
+                r.partition,
+                r.starting_offset,
+                r.max_partition_fetch_bytes,
+                r.ordinality AS idx -- for preserving original order
+            FROM unnest(arg_requests) WITH ORDINALITY AS r(topic_id, partition, starting_offset, max_partition_fetch_bytes, ordinality)
+        ),
+        requests_with_log_info AS (
+            SELECT
+                r.idx, r.topic_id, r.partition, r.starting_offset, r.max_partition_fetch_bytes,
+                l.log_start_offset, l.high_watermark, l.topic_name,
+                CASE
+                    WHEN l.topic_id IS NULL THEN 'unknown_topic_or_partition'::find_batches_response_error_v1
+                    WHEN r.starting_offset < 0 OR r.starting_offset > l.high_watermark THEN 'offset_out_of_range'::find_batches_response_error_v1
+                    ELSE NULL
+                END AS error
+            FROM requests r
+            LEFT JOIN logs l ON r.topic_id = l.topic_id AND r.partition = l.partition
+        ),
+        all_batches_with_metadata AS (
+            SELECT
+                r.idx,
+                (
+                    b.batch_id,
+                    f.object_key,
+                    (
+                        b.magic, b.topic_id, r.topic_name, b.partition, b.byte_offset, b.byte_size,
+                        b.base_offset, b.last_offset, b.log_append_timestamp, b.batch_max_timestamp,
+                        b.timestamp_type
+                    )::batch_metadata_v1
+                )::batch_info_v1 AS batch_data,
+                b.byte_size, b.base_offset, r.max_partition_fetch_bytes,
+                ROW_NUMBER() OVER (PARTITION BY r.idx ORDER BY b.base_offset) as rn,
+                SUM(b.byte_size) OVER (PARTITION BY r.idx ORDER BY b.base_offset) as partition_cumulative_bytes
+            FROM requests_with_log_info r
+            JOIN batches b ON r.topic_id = b.topic_id AND r.partition = b.partition
+            JOIN files f ON b.file_id = f.file_id
+            WHERE r.error IS NULL
+                AND b.last_offset >= r.starting_offset
+                AND b.base_offset < r.high_watermark
+        ),
+        per_partition_limited_batches AS (
+            SELECT idx, batch_data, byte_size, base_offset, rn
+            FROM all_batches_with_metadata
+            WHERE (rn = 1  -- each partition gets always at least one batch
+                -- include also last batch, even if it overflows max.partition.fetch.bytes
+                OR (partition_cumulative_bytes - byte_size) < max_partition_fetch_bytes
+                ) AND (max_batches_per_partition = 0 OR rn <= max_batches_per_partition)
```

Comment:
main change: same as before if new config == 0, otherwise, limit the number of partitions.